fix(22682): Fail on class declaration in statement position for ES2015 or higher

nicdard opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [x] There is an associated issue in the `Backlog` milestone (**required**)
* [x] Code is up-to-date with the `main` branch
* [x] You've successfully run `gulp runtests` locally
* [x] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/main/CONTRIBUTING.md
-->

Fixes #22682 

The error should be reported only in ES2015 or higher, therefore I changed the `binder` instead of the `parser` which needs still to parse this sequence to allow older standards.

In this PR I am checking only class declarations, since I am not sure from the conversation in the issue which is the required behaviour for enums.

Since this is my first contribution, I am not sure about the test coverage. In particular, I am not sure on how to generate the `.error.txt` file which I need to generate by running directly (calling tsc) the updated version of the compiler against the newly created `scopeClassDeclaration.ts` test file.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -42870,6 +42870,9 @@ namespace ts {
         }
 
         function checkGrammarClassLikeDeclaration(node: ClassLikeDeclaration): boolean {
+            if (languageVersion >= ScriptTarget.ES2015 && node.parent && isIfStatement(node.parent)) {
```
I don't think we want to gate this on the target.

nicdard replied:
Can you provide some more guidance? :)

nicdard replied:
Do you mean I should simply check always?

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1409,6 +1409,10 @@
         "category": "Error",
         "code": 1471
     },
+    "A class declaration cannot appear in statement-only position in 'ES2015' or higher." : {
```
maybe

```suggestion
    "A class declaration can only be used in a scope where other statements can reference it." : {
```

nicdard replied:
Ok!

DanielRosenwasser commented on tests/baselines/reference/scopeClassDeclaration.errors.txt:
```diff
@@ -0,0 +1,7 @@
+
+4     class C {} // runtime error if transpiled to ES2015 or above: Unexpected token class
+            ~
+
```
Hm, shouldn't there be an error message on this line? Did the way we do baselines change?

DanielRosenwasser replied:
> Since this is my first contribution, I am not sure about the test coverage. In particular, I am not sure on how to generate the `.error.txt` file which I need to generate by running directly

Ah, I see, you grabbed this from TypeScript's direct output. You should just run

```sh
gulp baseline-accept
```

add that to git, and let us review that.

nicdard replied:
Oh ok I see, thanks!

nicdard replied:
I think I have got this right.

The first time I created the error file manually since it was not created by the command, as I needed to run tsc with a specific configuration (due to the target check in `checker`)

By removing the check the error file has been created.

nicdard commented:
Hi I am sorry for the long time, I had examination session and some other stuff which took priority unfortunately :(

I will start working back on this tomorrow, thanks also for the review!

sandersn commented:
Do you want to keep working on this? Otherwise I'd like to close it to reduce the number of open PRs.

nicdard commented:
> Do you want to keep working on this? Otherwise I'd like to close it to reduce the number of open PRs.

Yes sure, however I didn't fully understand the feedback, especially the first one, can you provide more guidance?

nicdard commented:
@DanielRosenwasser I think I addressed all the issues you mentioned in your review (thanks again), still not entirely sure on the first suggestion, but in case let me know so I will fix it :)

jakebailey commented:
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test this

jakebailey commented on src/compiler/checker.ts:
```diff
@@ -43614,6 +43614,9 @@ namespace ts {
         }
 
         function checkGrammarClassLikeDeclaration(node: ClassLikeDeclaration): boolean {
+            if (node.parent && isIfStatement(node.parent)) {
```
Is this the right fix? I'm also wondering about cases like:

```ts
declare const condition: boolean;

if (condition) console.log(); else class X {}

function foo() {
	return class X {};
}
```

Which all should error (and maybe more). Maybe it's that the check should be checking if the parent is some sort of block?

jakebailey replied:
Also, I'm a little confused as to why this is in the checker; are these grammar errors that should be in the parser?

RyanCavanaugh replied:
Some grammar errors are computed in the check phase so that the parser can have better error recovery

RyanCavanaugh replied:
e.g. in this case, it's "better" for the parser to just accept the invalid class decl and then flag it later during a subsequent walk, as opposed to issuing individual errors for ~every token of the class

jakebailey replied:
I see, I was just guessing that it'd be like the other errors generated in the binder, e.g. the variable redeclaration stuff that was recently moved there to allow the errors to show in JS.

jakebailey replied:
> e.g. in this case, it's "better" for the parser to just accept the invalid class decl and then flag it later during a subsequent walk, as opposed to issuing individual errors for ~every token of the class

It'd do that? I would have guessed that it'd recover by parsing the class but emitting an error anyway along with the fully parsed class node.

nicdard replied:
Thanks for the review! I will implement also the other checks, should I move the checks or let them in the `checker`?

jakebailey replied:
You should be able to just leave them in the `checkGrammar` functions (and assume my intuition here is wrong).

nicdard replied:
Ok thanks, then I will add other checks there to also avoid cases that you pointed out :)

jakebailey replied:
I sort of misspoke, it'd still be in the function you just edited because that's the one that's checking a class declaration, it's just that you need to check for other wrong parent node types. Or, the other way around in that maybe the fix is to error if the parent is not a block or something.

nicdard replied:
Thanks for the clarification!

I was also wondering on something like this:
```
while (class C {}) {
    break;
}
```
this should also be included right?

Also I noticed that all the cases we are mentioning, for example:
```
function foo() {
    console.log(12);
	return class X {};
}
const A = foo()
```
Is actually possible now and works in js, so this would be a breaking change, right?

jakebailey replied:
Probably, if it doesn't have an error already. This supports my "parent is not a block" rule.

For the second case, I think I tested that and it failed. Are you seeing it work?

nicdard replied:
The last test seems  to work now, I checked also with the [last release](https://www.typescriptlang.org/play?ts=4.6.2#code/GYVwdgxgLglg9mABMOcAUBKRBvAUIgxCBAZzgBsBTAOnLgHM0BGAJgwG5cBIAJ0qhA8kEcgEMSJRAA0cAX06zcxMCSiIAgogC8yVJiA) in the playground and with the [nightly build](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220406#code/GYVwdgxgLglg9mABMOcAUBKRBvAUIgxCBAZzgBsBTAOnLgHM0BGAJgwG5cBIAJ0qhA8kEcgEMSJRAA0cAX06zcxMCSiIAgogC8yVJiA)

nicdard replied:
The parent is a block rule makes absolute sense yes, thanks!

nicdard replied:
I implemented the fix with the block check :)

jakebailey replied:
Cool, thanks!

nicdard replied:
mmm I am seeing this failing some compiler tests (sorry for this, my own computer is really to slow so I usually push to use the github infrastructure, I have tried the one test I modified but not the rest)

jakebailey reviewed (changes requested):
Forgot to hit the request changes button. See above for comments.

jakebailey commented:
@typescript-bot user test this inline

jakebailey commented:
Hm, this fails quite a bit; I don't think you ran the tests before sending.

I think there's more things to check than just `isBlock` (there's module blocks, case blocks, or even the source file itself). There's a helper in `extractSymbol` called `isBlockLike` that would be a good candidate for moving into `src/compiler/utilities.ts` as it appears to do what's intended.

nicdard commented:
> Hm, this fails quite a bit; I don't think you ran the tests before sending.
> 
> I think there's more things to check than just `isBlock` (there's module blocks, case blocks, or even the source file itself). There's a helper in `extractSymbol` called `isBlockLike` that would be a good candidate for moving into `src/compiler/utilities.ts` as it appears to do what's intended.

Yeah sorry for this, I was warning about in the thread: 
mmm I am seeing this failing some compiler tests (sorry for this, my own computer is really to slow so I usually push to use the github infrastructure, I have tried the one test I modified but not the rest)

Oh cool thanks for the hint!

nicdard commented:
> Hm, this fails quite a bit; I don't think you ran the tests before sending.
> 
> I think there's more things to check than just `isBlock` (there's module blocks, case blocks, or even the source file itself). There's a helper in `extractSymbol` called `isBlockLike` that would be a good candidate for moving into `src/compiler/utilities.ts` as it appears to do what's intended.

Actually I have taken a better look at the tests that are failing, and there are cases such:
```
/// anonClassDeclarationEmitsAnon.ts

// @declaration: true
// @filename: wrapClass.ts
export function wrapClass(param: any) {
    return class Wrapped {
        foo() {
            return param;
        }
    }
}
```

or

```
/// classExpressionWithStaticPropertiesES61.ts

//@target: es6
var v = class C { 
    static a = 1;
    static b = 2;
    static c = C.a + 3;
};
```
I think we should test first if the node is a class statement then, so we avoid all those errors on class expressions, am I on the right track?

jakebailey commented:
Clearly I was wrong about the rules, then. This code executes today: [Playground Link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAdwE4EMAOBhANugZwIApN0MBbALkXTAE8BKRAbwChFPFUBTKEVEgj4iiAOoZMmHgBNWHLouBw4xZu0WauvfoMRlKAbgVaAvicTnTQA)

Probably worth testing all of the cases and see what's actually disallowed?

nicdard commented:
> Clearly I was wrong about the rules, then. This code executes today: [Playground Link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAdwE4EMAOBhANugZwIApN0MBbALkXTAE8BKRAbwChFPFUBTKEVEgj4iiAOoZMmHgBNWHLouBw4xZu0WauvfoMRlKAbgVaAvicTnTQA)
> 
> Probably worth testing all of the cases and see what's actually disallowed?

Maybe the intution was ok, but just regarding `classDeclaration`, I'll try, otherwise will go through all cases

jakebailey commented on src/compiler/checker.ts:
```diff
@@ -39276,13 +39276,34 @@ namespace ts {
             registerForUnusedIdentifiersCheck(node);
         }
 
+        function isBlockLike(node: Node): node is BlockLike {
+            switch (node.kind) {
+                case SyntaxKind.Block:
+                case SyntaxKind.SourceFile:
+                case SyntaxKind.ModuleBlock:
+                case SyntaxKind.CaseClause:
+                    return true;
+                default:
+                    return false;
+            }
+        }
```
This should probably have been moved from `extractSymbol.ts` into `src/compiler/utils.ts` rather than being copied and pasted here.

nicdard replied:
Yeah I was trying the tests :) Sorry for this, it's just my computer really cannot run the tests without exploding :sweat_smile: Is there any tip to run them locally without overwhelming it?

jakebailey commented on tests/cases/compiler/scopeClassDeclarations.ts:
```diff
@@ -0,0 +1,19 @@
+
+while (class C {}) {  // Not an error.
+    break;
+}
```
This file should also test the other missing parents, e.g. case, default, etc.

nicdard replied:
Oh sure thanks!

nicdard commented:
I'll change this to draft so I can work on it without disturbing, as soon as I think it is ready I will reopen it

sandersn commented:
This PR is quite old now. I'm going to close it for now and we can re-open it if needed.
